### ISSUE TITLE
Combining flags

### ISSUE THREAD

**Issue by DieracDelta** [draperlaboratory/cbat_tools#137]
In the following example, the combination of ` --wp-check-null-deref=true --wp-check-calls=true` when invoking CBAT results in refuted non-null load and non-null store goal as expected. Given that all reads in `main_1` are valid, `main_1` does not dereference `%rax` when `%rax` is `NULL`, but `main_2` however does dereference `%rax` in all executions.

When calling with `--wp-check-calls=true` the result is `UNSAT`, and for `--wp-check-null-deref=true` the result is a refuted goal of matching return addresses. When `CBAT` is invoked with neither of the flags, the refuted goal is also matching return addresses. The expected behavior is the `--wp-check-null-deref=true` flag would catch the null dereference, and the `--wp-check-calls` flag to not have an effect.

The assembly for `main_1.S`:
```
.text
x2apic_dead_cpu:
	mov    0x0(,%rdi,8),%rdx
	mov    (%rdx),%rax
	test   %rax,%rax
	je finish
	movq $0x0,(%rax)
finish:
	xor    %eax,%eax
	retq
```
The BIR for main_1:
```
0000007f: program
00000078: sub finish()
00000037: 
0000004d: RAX := 0
00000050: AF := unknown[bits]:u1
00000053: ZF := 1
00000056: PF := 1
00000059: OF := 0
0000005c: CF := 0
0000005f: SF := 0
00000068: #3 := mem[RSP, el]:u64
0000006b: RSP := RSP + 8
0000006f: call #3 with noreturn

00000079: sub x2apic_dead_cpu()
00000007: 
0000000c: RDX := mem[RDI << 3, el]:u64
00000014: RAX := mem[RDX, el]:u64
00000021: #1 := RAX
00000024: OF := 0
00000027: CF := 0
0000002a: AF := unknown[bits]:u1
0000002d: PF := ~low:1[let $1 = #1 >> 4 ^ #1 in let $2 = $1 >> 2 ^ $1 in
$2 >> 1 ^ $2]
00000030: SF := high:1[#1]
00000033: ZF := 0 = #1
0000003d: when ZF call @finish with noreturn
0000007d: goto %00000071

00000071: 
00000076: mem := mem with [RAX, el]:u64 <- 0
0000007e: call @finish with noreturn
```

The assembly for `main_2.S`:
```
.text
x2apic_dead_cpu:
	mov    0x0(,%rdi,8),%rdx
	mov    (%rdx),%rax
	movq $0x0,(%rax)
	xor    %eax,%eax
	retq
```
The BIR for `main_2`:
```
0000004e: program
0000004c: sub x2apic_dead_cpu()
00000007: 
0000000c: RDX := mem[RDI << 3, el]:u64
00000014: RAX := mem[RDX, el]:u64
0000001b: mem := mem with [RAX, el]:u64 <- 0
00000028: RAX := 0
0000002b: AF := unknown[bits]:u1
0000002e: ZF := 1
00000031: PF := 1
00000034: OF := 0
00000037: CF := 0
0000003a: SF := 0
00000043: #1 := mem[RSP, el]:u64
00000046: RSP := RSP + 8
0000004a: call #1 with noreturn
```

The invocation for compiling and running with CBAT:
```
gcc -shared -c -fno-stack-protector -Wall -o main_1 main_1.S
gcc -shared -c -fno-stack-protector -Wall -o main_2 main_2.S
bap main_1 --pass=save-project --save-project-filename=main_1.bpj
bap main_2 --pass=save-project --save-project-filename=main_2.bpj
bap /bin/echo --pass=wp \
    --wp-compare=true \
    --wp-file1=main_1.bpj \
    --wp-file2=main_2.bpj \
    --wp-function=x2apic_dead_cpu \
    --wp-check-null-deref=true\
    --wp-check-calls=true
```

**Comment by fortunac** [comment]
Looks like this behavior is due to the hypotheses used based on the flags. The default behavior of `wp` has checks the property that if every register between the two binaries are equal at the beginning of execution, then the output register(s) should be equal at the end of execution.

If `check_calls` is set, the property stated above is not checked, but rather we check if the registers between the binaries are equal at the beginning of execution and a function has not been called in the original binary, then it won't be called in the modified binary.

These two properties we check are disjoint from each other, and we currently don't check both at the same time.

In our register comparison case, we have two hypotheses that seem to be preventing us from catching the `null_deref` checks. These hypotheses seem to be absent in our check function calls case.
- Constraining the initial value of RSP to be on the stack
    - Constraining RSP to be on the stack looks like its the correct behavior. One of the VCs that was refuted stated that RSP was 0x0, and we were reading the stack at that point, which should not be the case. We should probably include this hypothesis in the function comparison case.
- Generating variables that represent the initial value of a register and creating constraints that enforce this (i.e. `init_RDI0 == RDI0`)
    - I think this is a bug. I will have to look further into why this hypothesis prevents us from catching those null checks.

**Comment by fortunac** [comment]
Closed with #143 and #158.
